### PR TITLE
CRI: verify responses from remote runtime

### LINKED ISSUE #40264: CRI: verify responses from remote runtime
CRI doesn't verify responses from remote runtime today. We should verify and check all required responses so as to

- fail fast, e.g. by verifying imageRef for PullImage, kubelet could fail at image pulling stage instead of syncPod
- easy debugging
- protect kubelet 

cc @yujuhong @Random-Liu @yifan-gu

### PR DESCRIPTION
**Pull request by feiskyer**
Closes #40264.

### CODE REVIEW

**Comment by feiskyer** [comment]
@k8s-bot kops aws e2e test this

File: pkg/kubelet/remote/remote_runtime.go [hunk 1]
```diff
@@ -129,6 +143,13 @@ func (r *RemoteRuntimeService) PodSandboxStatus(podSandBoxID string) (*runtimeap
 		return nil, err
 	}
 
+	if resp.Status != nil {
+		if err := verifySandboxStatus(resp.Status); err != nil {
+			glog.Errorf("PodSandboxStatus of %q failed: %v", podSandBoxID, err)
```
Comment (yifan-gu) [inline pkg/kubelet/remote/remote_runtime.go]: nit: Have we already logged errors in the caller?

File: pkg/kubelet/remote/utils.go [hunk 2]
```diff
@@ -32,3 +35,59 @@ func dial(addr string, timeout time.Duration) (net.Conn, error) {
...
+	metadata := status.Metadata
+	if metadata.Name == "" || metadata.Namespace == "" || metadata.Uid == "" {
+		return fmt.Errorf("Name, Namespace or Uid is not in metadata %q", metadata)
+	}
+
+	if status.CreatedAt == 0 {
+		return fmt.Errorf("CreatedAt is not set")
+	}
+
+	if status.Network == nil {
```
Comment (yifan-gu) [inline pkg/kubelet/remote/utils.go]: Unready pod sandbox can have empty network, no?

File: pkg/kubelet/remote/utils.go [hunk 3]
```diff
@@ -32,3 +35,59 @@ func dial(addr string, timeout time.Duration) (net.Conn, error) {
...
+	}
+
+	if status.Metadata == nil {
+		return fmt.Errorf("Metadata is not set")
+	}
+
+	metadata := status.Metadata
+	if metadata.Name == "" {
+		return fmt.Errorf("Name is not in metadata %q", metadata)
+	}
```
Comment (yifan-gu) [inline pkg/kubelet/remote/utils.go]: nit: (optionally) check the attemp is >= 0.

File: pkg/kubelet/remote/utils.go [hunk 4]
```diff
@@ -32,3 +35,59 @@ func dial(addr string, timeout time.Duration) (net.Conn, error) {
...
+
+	if status.CreatedAt == 0 {
+		return fmt.Errorf("CreatedAt is not set")
+	}
+
+	if status.Image == nil {
+		return fmt.Errorf("Image is not set")
+	}
```
Comment (yifan-gu) [inline pkg/kubelet/remote/utils.go]: nit: Do we need to check `status.Image.Image ! = ""` as well?

**Comment by yifan-gu** [comment]
Thanks @feiskyer , this is very useful for runtime developers.
Couple nits, lgtm otherwise.

File: pkg/kubelet/remote/remote_runtime.go (hunk 1 above)
Comment (feiskyer) [inline pkg/kubelet/remote/remote_runtime.go]: yes, should delete this log message here.

File: pkg/kubelet/remote/utils.go (hunk 3 above)
Comment (feiskyer) [inline pkg/kubelet/remote/utils.go]: it is already `uint32`

File: pkg/kubelet/remote/utils.go (hunk 4 above)
Comment (feiskyer) [inline pkg/kubelet/remote/utils.go]: ack

File: pkg/kubelet/remote/utils.go (hunk 2 above)
Comment (feiskyer) [inline pkg/kubelet/remote/utils.go]: ack

**Comment by feiskyer** [comment]
@yifan-gu Addressed comments. PTAL.

**Comment by feiskyer** [comment]
cc @kubernetes/sig-node-pr-reviews

**Comment by feiskyer** [comment]
@k8s-bot test this

**Pull request merged**